File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/history/TimelineEntry.java
package eu.pkgsoftware.babybuddywidgets.history;

import android.view.View;

import com.squareup.phrase.Phrase;

import java.net.MalformedURLException;
import java.text.DateFormat;
import java.util.Date;
import java.util.Objects;

import eu.pkgsoftware.babybuddywidgets.BaseFragment;
import eu.pkgsoftware.babybuddywidgets.Constants;
import eu.pkgsoftware.babybuddywidgets.R;
import eu.pkgsoftware.babybuddywidgets.databinding.TimelineItemBinding;
import eu.pkgsoftware.babybuddywidgets.networking.BabyBuddyClient;
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.ChangeEntry;
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.FeedingEntry;
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.TimeEntry;

public class TimelineEntry {
    public static final DateFormat DATE_FORMAT = DateFormat.getDateInstance(DateFormat.SHORT);
    public static final DateFormat TIME_FORMAT = DateFormat.getTimeInstance(DateFormat.SHORT);

    private final BaseFragment fragment;
    private final TimelineItemBinding binding;

    private TimeEntry entry = null;

    private Runnable modifiedCallback = null;

    private void hideAllSubviews() {
        for (int i = 0; i < binding.viewGroup.getChildCount(); i++) {
            View c = binding.viewGroup.getChildAt(i);
            c.setVisibility(View.GONE);
        }
    }

    private Phrase defaultPhraseFields(Phrase phrase) {
        return phrase
            .putOptional("type", entry.getType())
            .putOptional("start_date", DATE_FORMAT.format(entry.getStart()))
            .putOptional("start_time", TIME_FORMAT.format(entry.getStart()))
            .putOptional("end_date", DATE_FORMAT.format(entry.getEnd()))
            .putOptional("end_time", TIME_FORMAT.format(entry.getEnd()))
            .putOptional("notes", entry.getNotes().trim());
    }

    private void configureDefaultView() {
        hideAllSubviews();
        binding.viewGroup.getChildAt(0).setVisibility(View.VISIBLE);

        String message = defaultPhraseFields(
            Phrase.from("{type}\n{start_date}  {start_time} - {end_time}")
        ).format().toString();

        binding.defaultContent.setText(message);
    }

    private void configureTummyTime() {
        hideAllSubviews();
        binding.tummyTimeView.setVisibility(View.VISIBLE);

        String message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time} - {end_time}\n{notes}")
        ).format().toString().trim();

        binding.tummytimeMilestoneText.setText(message);
    }

    private void configureChange() {
        hideAllSubviews();
        binding.diaperView.setVisibility(View.VISIBLE);

        ChangeEntry change = (ChangeEntry) entry;
        binding.diaperWetImage.setVisibility(change.getWet() ? View.VISIBLE : View.GONE);
        binding.diaperSolidImage.setVisibility(change.getSolid() ? View.VISIBLE : View.GONE);

        String message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time}\n{notes}")
        ).format().toString().trim();

        binding.diaperText.setText(message.trim());
    }

    private void configureSleep() {
        hideAllSubviews();
        binding.sleepView.setVisibility(View.VISIBLE);

        String message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time} - {end_time}\n{notes}")
        ).format().toString().trim();

        binding.sleepText.setText(message.trim());
    }

    private void configureNote() {
        hideAllSubviews();
        binding.noteTimeView.setVisibility(View.VISIBLE);

        String message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time}\n{notes}")
        ).format().toString().trim();

        binding.noteTimeEntryText.setText(message.trim());
    }

    private void configurePumping() {
        hideAllSubviews();
        binding.pumpingTimeView.setVisibility(View.VISIBLE);

        String message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time} - {end_time}\n{notes}")
        ).format().toString().trim();

        binding.pumpingTimeNotes.setText(message.trim());
    }

    private void configureFeeding() {
        hideAllSubviews();
        binding.feedingView.setVisibility(View.VISIBLE);

        FeedingEntry feeding = (FeedingEntry) entry;

        binding.feedingBreastImage.setVisibility(View.GONE);
        binding.feedingBreastLeftImage.setVisibility(View.GONE);
        binding.feedingBreastRightImage.setVisibility(View.GONE);
        binding.feedingBottleImage.setVisibility(View.GONE);
        binding.solidFoodImage.setVisibility(View.GONE);

        switch (Constants.FeedingTypeEnum.byPostName(feeding.getFeedingType())) {
            case BREAST_MILK:
                final Constants.FeedingMethodEnum feedingMethod =
                    Constants.FeedingMethodEnum.byPostName(feeding.getFeedingMethod());
                if (feedingMethod.value == 1) {
                    binding.feedingBreastLeftImage.setVisibility(View.VISIBLE);
                    break;
                } else if (feedingMethod.value == 2) {
                    binding.feedingBreastRightImage.setVisibility(View.VISIBLE);
                    break;
                } else if (feedingMethod.value == 3) {
                    binding.feedingBreastImage.setVisibility(View.VISIBLE);
                    break;
                }
            case FORTIFIED_BREAST_MILK:
            case FORMULA:
                binding.feedingBottleImage.setVisibility(View.VISIBLE);
                break;
            case SOLID_FOOD:
                binding.solidFoodImage.setVisibility(View.VISIBLE);
                break;
            default:
                binding.solidFoodImage.setVisibility(View.VISIBLE);
        }

        String message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time} - {end_time}\n{notes}")
        ).format().toString().trim();

        binding.feedingText.setText(message.trim());
    }

    private boolean longClick() {
        if (entry != null) {
            BabyBuddyClient client = fragment.getMainActivity().getClient();
            try {
                fragment.showUrlInBrowser(client.v2client.entryUserPath(entry).toString());
                return true;
            } catch (MalformedURLException e) {
                e.printStackTrace();
                return false;
            }
        }
        return false;
    }

    private void removeClick() {
        if (entry == null) {
            return;
        }
        final TimeEntry thisEntry = entry;

        fragment.showQuestion(
            true,
            fragment.getResources().getString(R.string.history_delete_title),
            defaultPhraseFields(
                Phrase.from(fragment.getMainActivity(), R.string.history_delete_question)
            ).format().toString().trim(),
            fragment.getResources().getString(R.string.history_delete_question_delete_button),
            fragment.getResources().getString(R.string.history_delete_question_cancel_button),
            b -> {
                if (!b) {
                    return;
                }
                BabyBuddyClient client = fragment.getMainActivity().getClient();
                /* TODO  client.removeTimelineEntry(thisEntry, new BabyBuddyClient.RequestCallback<Boolean>() {
                    @Override
                    public void error(@NotNull Exception error) {
                    }

                    @Override
                    public void response(Boolean response) {
                        setTimeEntry(null);
                        if (modifiedCallback != null) {
                            modifiedCallback.run();
                        }
                    }
                });*/
            }
        );
    }

    public TimelineEntry(BaseFragment fragment, TimeEntry entry) {
        this.fragment = fragment;

        binding = TimelineItemBinding.inflate(fragment.getMainActivity().getLayoutInflater());
        setTimeEntry(entry);

        binding.getRoot().setOnLongClickListener((v) -> longClick());
        binding.removeButton.setOnClickListener((v) -> removeClick());
    }

    public void setTimeEntry(TimeEntry entry) {
        if (Objects.equals(this.entry, entry)) {
            return;
        }
        this.entry = entry;

        if (entry == null) {
            binding.getRoot().setVisibility(View.INVISIBLE);
        } else {
            binding.getRoot().setVisibility(View.VISIBLE);

            if (BabyBuddyClient.ACTIVITIES.TUMMY_TIME.equals(entry.getType())) {
                configureTummyTime();
            } else if (BabyBuddyClient.EVENTS.CHANGE.equals(entry.getType())) {
                configureChange();
            } else if (BabyBuddyClient.ACTIVITIES.SLEEP.equals(entry.getType())) {
                configureSleep();
            } else if (BabyBuddyClient.ACTIVITIES.FEEDING.equals(entry.getType())) {
                configureFeeding();
            } else if (BabyBuddyClient.EVENTS.NOTE.equals(entry.getType())) {
                configureNote();
            } else if (BabyBuddyClient.ACTIVITIES.PUMPING.equals(entry.getType())) {
                configurePumping();
            } else {
                configureDefaultView();
            }
        }
    }

    public TimeEntry getTimeEntry() {
        return entry;
    }

    public View getView() {
        return binding.getRoot();
    }

    public Date getDate() {
        return entry.getEnd();
    }

    public void setModifiedCallback(Runnable r) {
        modifiedCallback = r;
    }
}

File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/history/TimelineEntry.kt
package eu.pkgsoftware.babybuddywidgets.history

import android.view.MotionEvent
import android.view.View
import com.squareup.phrase.Phrase
import eu.pkgsoftware.babybuddywidgets.BaseFragment
import eu.pkgsoftware.babybuddywidgets.Constants
import eu.pkgsoftware.babybuddywidgets.Constants.FeedingMethodEnum
import eu.pkgsoftware.babybuddywidgets.Constants.FeedingTypeEnum
import eu.pkgsoftware.babybuddywidgets.DialogCallback
import eu.pkgsoftware.babybuddywidgets.R
import eu.pkgsoftware.babybuddywidgets.databinding.TimelineItemBinding
import eu.pkgsoftware.babybuddywidgets.networking.BabyBuddyClient
import eu.pkgsoftware.babybuddywidgets.networking.RequestCodeFailure
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.ChangeEntry
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.FeedingEntry
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.models.TimeEntry
import eu.pkgsoftware.babybuddywidgets.networking.babybuddy.serverTimeToClientTime
import kotlinx.coroutines.launch
import java.io.IOException
import java.net.MalformedURLException
import java.text.DateFormat


class TimelineEntry(private val fragment: BaseFragment, private var _entry: TimeEntry?) {
    companion object {
        val DATE_FORMAT = DateFormat.getDateInstance(DateFormat.SHORT)
        val TIME_FORMAT = DateFormat.getTimeInstance(DateFormat.SHORT)
    }

    private val binding = TimelineItemBinding.inflate(fragment.mainActivity.layoutInflater)
    private var modifiedCallback: Runnable? = null
    private fun hideAllSubviews() {
        for (i in 0 until binding.viewGroup.childCount) {
            val c = binding.viewGroup.getChildAt(i)
            c.visibility = View.GONE
        }
    }

    var entry: TimeEntry?
        get() = _entry
        set(v) {
            if (v == _entry) {
                return
            }
            _entry = v
            updateUi()
        }

    val view: View
        get() = binding.root

    private fun updateUi() {
        val entry = _entry
        if (entry == null) {
            binding.root.visibility = View.INVISIBLE
        } else {
            binding.root.visibility = View.VISIBLE
            if (BabyBuddyClient.ACTIVITIES.TUMMY_TIME == entry.appType) {
                configureTummyTime()
            } else if (BabyBuddyClient.EVENTS.CHANGE == entry.appType) {
                configureChange()
            } else if (BabyBuddyClient.ACTIVITIES.SLEEP == entry.appType) {
                configureSleep()
            } else if (BabyBuddyClient.ACTIVITIES.FEEDING == entry.appType) {
                configureFeeding()
            } else if (BabyBuddyClient.EVENTS.NOTE == entry.appType) {
                configureNote()
            } else if (BabyBuddyClient.ACTIVITIES.PUMPING == entry.appType) {
                configurePumping()
            } else {
                configureDefaultView()
            }
        }
    }

    init {
        binding.root.setOnLongClickListener { v: View? -> longClick() }
        binding.root.setOnTouchListener { v, event ->
            longClickStartStopHandler(v, event)
            false
        }
        binding.removeButton.setOnClickListener { v: View? -> removeClick() }
        updateUi()
    }

    private fun defaultPhraseFields(phrase: Phrase): Phrase {
        val local_start_time = serverTimeToClientTime(entry!!.start)
        val local_end_time = serverTimeToClientTime(entry!!.end)

        val start_time = TIME_FORMAT.format(local_start_time)
        val end_time = TIME_FORMAT.format(local_end_time)
        val opt_time_range = if (start_time == end_time) start_time else "$start_time - $end_time"

        return phrase
            .putOptional("type", entry!!.appType)
            .putOptional("start_date", DATE_FORMAT.format(local_start_time))
            .putOptional("start_time", TIME_FORMAT.format(local_start_time))
            .putOptional("end_date", DATE_FORMAT.format(local_end_time))
            .putOptional("end_time", TIME_FORMAT.format(local_end_time))
            .putOptional("opt_time_range", opt_time_range)
            .putOptional("notes", entry!!.notes.trim { it <= ' ' })
    }

    private fun configureDefaultView() {
        hideAllSubviews()
        binding.viewGroup.getChildAt(0).visibility = View.VISIBLE
        val message = defaultPhraseFields(
            Phrase.from("{type}\n{start_date}  {opt_time_range}")
        ).format().toString()
        binding.defaultContent.text = message
    }

    private fun configureTummyTime() {
        hideAllSubviews()
        binding.tummyTimeView.visibility = View.VISIBLE
        val message = defaultPhraseFields(
            Phrase.from("{start_date}  {opt_time_range}\n{notes}")
        ).format().toString().trim { it <= ' ' }
        binding.tummytimeMilestoneText.text = message
    }

    private fun configureChange() {
        hideAllSubviews()
        var amountString = ""

        binding.diaperView.visibility = View.VISIBLE
        (entry as ChangeEntry?)?.let { change ->
            binding.diaperWetImage.visibility =
                if (change.wet) View.VISIBLE else View.GONE
            binding.diaperSolidImage.visibility = if (change.solid) View.VISIBLE else View.GONE

            if (change.color.isNotEmpty()) {
                binding.diaperColorPreview.visibility = View.VISIBLE

                val colorEnumValue = Constants.SolidDiaperColorEnum.byPostName(change.color)
                fragment.resources.getColor(colorEnumValue.colorResId, null).let { color ->
                    binding.diaperColorPreview.setBackgroundColor(color)
                }
            } else {
                binding.diaperColorPreview.visibility = View.GONE
            }

            if (change.amount != null) {
                amountString =
                    Phrase.from(fragment.resources, R.string.diaper_amount_timeline_pattern)
                        .put("amount", String.format("%.3f", change.amount))
                    .format().toString()
                amountString += "\n"
            }
        }
        val message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time}\n{amount}{notes}")
        ).put("amount", amountString).format().toString().trim { it <= ' ' }
        binding.diaperText.text = message.trim { it <= ' ' }
    }

    private fun configureSleep() {
        hideAllSubviews()
        binding.sleepView.visibility = View.VISIBLE
        val message = defaultPhraseFields(
            Phrase.from("{start_date}  {opt_time_range}\n{notes}")
        ).format().toString().trim { it <= ' ' }
        binding.sleepText.text = message.trim { it <= ' ' }
    }

    private fun configureNote() {
        hideAllSubviews()
        binding.noteTimeView.visibility = View.VISIBLE
        val message = defaultPhraseFields(
            Phrase.from("{start_date}  {start_time}\n{notes}")
        ).format().toString().trim { it <= ' ' }
        binding.noteTimeEntryText.text = message.trim { it <= ' ' }
    }

    private fun configurePumping() {
        hideAllSubviews()
        binding.pumpingTimeView.visibility = View.VISIBLE
        val message = defaultPhraseFields(
            Phrase.from("{start_date}  {opt_time_range}\n{notes}")
        ).format().toString().trim { it <= ' ' }
        binding.pumpingTimeNotes.text = message.trim { it <= ' ' }
    }

    private fun configureFeeding() {
        hideAllSubviews()
        binding.feedingView.visibility = View.VISIBLE
        val feeding = entry!! as FeedingEntry
        binding.feedingBreastImage.visibility = View.GONE
        binding.feedingBreastLeftImage.visibility = View.GONE
        binding.feedingBreastRightImage.visibility = View.GONE
        binding.feedingBottleImage.visibility = View.GONE
        binding.solidFoodImage.visibility = View.GONE
        when (FeedingTypeEnum.byPostName(feeding.feedingType)) {
            FeedingTypeEnum.BREAST_MILK -> {
                val feedingMethod = FeedingMethodEnum.byPostName(
                    feeding.feedingMethod
                )
                if (feedingMethod.value == 1) {
                    binding.feedingBreastLeftImage.visibility = View.VISIBLE
                } else if (feedingMethod.value == 2) {
                    binding.feedingBreastRightImage.visibility = View.VISIBLE
                } else if (feedingMethod.value == 3) {
                    binding.feedingBreastImage.visibility = View.VISIBLE
                } else {
                    binding.feedingBottleImage.visibility = View.VISIBLE
                }
            }

            FeedingTypeEnum.FORTIFIED_BREAST_MILK, FeedingTypeEnum.FORMULA -> binding.feedingBottleImage.visibility =
                View.VISIBLE

            FeedingTypeEnum.SOLID_FOOD -> binding.solidFoodImage.visibility = View.VISIBLE
            else -> binding.solidFoodImage.visibility = View.VISIBLE
        }
        val message = defaultPhraseFields(
            Phrase.from("{start_date}  {opt_time_range}\n{notes}")
        ).format().toString().trim { it <= ' ' }
        binding.feedingText.text = message.trim { it <= ' ' }
    }

    private fun longClickStartStopHandler(v: View, event: MotionEvent) {
        if (event.action == MotionEvent.ACTION_DOWN) {
            binding.longclickBubble.startGrow()
        } else if (event.action in listOf(
                MotionEvent.ACTION_UP,
                MotionEvent.ACTION_CANCEL,
                MotionEvent.ACTION_OUTSIDE,
            )
        ) {
            binding.longclickBubble.stopGrow()
        }
    }

    private fun longClick(): Boolean {
        binding.longclickBubble.stopGrow()
        val thisEntry = entry ?: return false
        val client = fragment.mainActivity.client.v2client
        return try {
            fragment.showUrlInBrowser(client.entryUserPath(thisEntry).toString())
            true
        } catch (e: MalformedURLException) {
            e.printStackTrace()
            false
        }
    }

    private fun removeClick() {
        val thisEntry: TimeEntry = entry ?: return
        fragment.showQuestion(
            true,
            fragment.resources.getString(R.string.history_delete_title),
            defaultPhraseFields(
                Phrase.from(fragment.mainActivity, R.string.history_delete_question)
            ).format().toString().trim { it <= ' ' },
            fragment.resources.getString(R.string.history_delete_question_delete_button),
            fragment.resources.getString(R.string.history_delete_question_cancel_button),
            object : DialogCallback {
                override fun call(b: Boolean) {
                    if (!b) {
                        return
                    }
                    val client = fragment.mainActivity.client
                    fragment.mainActivity.scope.launch {
                        try {
                            client.v2client.deleteEntry(thisEntry)
                            this@TimelineEntry.entry = null
                            modifiedCallback?.run()
                        } catch (e: RequestCodeFailure) {
                            e.printStackTrace()
                        } catch (e: IOException) {
                            e.printStackTrace()
                        }
                    }                }
            }
        )
    }

    fun setModifiedCallback(r: Runnable?) {
        modifiedCallback = r
    }
}